### PR TITLE
added smart motor controller drivetrain

### PR DESCRIPTION
**Pull request by Yoel-Perman-Brilliant**

### CODE REVIEW

**Comment by Yoel-Perman-Brilliant** [comment]
no trapezim yet

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+     * @return {@code true} if the command should finish running, {@code false} otherwise
+     */
+    @Override
+    public boolean isFinished() {
+        if (!subsystem.onTarget(controlMode, leftPIDSettings.getTolerance(), rightPIDSettings.getTolerance(),
+                leftSetpoint.get(), rightSetpoint.get())) {
+            lastTimeLeftNotOnTarget = Timer.getFPGATimestamp();
+            lastTimeRightNotOnTarget = Timer.getFPGATimestamp();
+        }
+        //is this good? it seems wrong
```
Comment (OfriRosenbaum) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:132]: it's fine but you are calling the `getFPGATimestamp` many times, maybe save this as a variable at the start of the function (`time = Timer.getFPGATimestamp` and use time)

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    private double lastTimeLeftNotOnTarget;
+
+    /**
+     * The most recent timestamp on which the right side's loop has not reached its target setpoint.
+     */
+    private double lastTimeRightNotOnTarget;
+
+    /**
+     * Constructs a new (generic) instance of {@link MoveSmartMotorControllerTankDrivetrain}.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:67]: remove generic

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+     *
+     * @param drivetrain          the {@link SmartMotorControllerGenericSubsystem} this command will run on
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:69]: drivetrain

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+     * @param leftPIDSettings     the left side's loop's PID constants
+     * @param rightPIDSettings    the right side's loop's PID constants
+     * @param feedForwardSettings the loop's feed forward gains
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:72]: loops'

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+     * @param controlMode         the loop's control mode (e.g. voltage, velocity, position...)
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:73]: loops'

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+     * @param leftSetpoint        the setpoint this command should bring the
+     *                            {@link SmartMotorControllerGenericSubsystem}'s left side to
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:75]: drivetrain

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+     * @param rightSetpoint       the setpoint this command should bring the
+     *                            {@link SmartMotorControllerGenericSubsystem}'s right side to
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:77]: ditto

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * Configures the drivetrain's control loops.
+     */
+    @Override
+    public void initialize() {
+        drivetrain.configureLoop(leftPIDSettings, rightPIDSettings, feedForwardSettings);
+    }
+
+    /**
+     * Updates any control loops running on the drivetrain.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrain.java:105]: drivetrain's motor controllers

File: src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrainTrapezically.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import com.spikes2212.command.drivetrains.SmartMotorControllerTankDrivetrain;
+import com.spikes2212.control.FeedForwardSettings;
+import com.spikes2212.control.PIDSettings;
+import com.spikes2212.control.TrapezoidProfileSettings;
+import com.spikes2212.util.UnifiedControlMode;
+
+import java.util.function.Supplier;
+
+/**
+ * A command that moves a {@link SmartMotorControllerTankDrivetrain} by running a trapezoid profile on its
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrainTrapezically.java:12]: by running trapezoid profile control loops on its master motor controllers

Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/commands/MoveSmartMotorControllerTankDrivetrainTrapezically.java:12]: by running trapezoid profile control loops on its master motor controllers

File: src/main/java/com/spikes2212/command/genericsubsystem/commands/smartmotorcontrollersubsystem/MoveSmartMotorControllerSubsystem.java [hunk 10]
```diff
@@ -10,17 +10,17 @@
...
 
 /**
- * A command that moves a {@link SmartMotorControllerSubsystem} using its master motor controller's control loops.
+ * A command that moves a {@link SmartMotorControllerGenericSubsystem} using its master motor controller's control loops.
  *
  * @author Yoel Perman Brilliant
- * @see SmartMotorControllerSubsystem
+ * @see SmartMotorControllerGenericSubsystem
  */
 public class MoveSmartMotorControllerSubsystem extends CommandBase {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/genericsubsystem/commands/smartmotorcontrollersubsystem/MoveSmartMotorControllerSubsystem.java:18]: Generic

File: src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/CTRESmartMotorControllerGenericSubsystem.java [hunk 11]
```diff
@@ -19,15 +19,14 @@
  * A {@link Subsystem} which consists of a master CTRE motor controller that can run control loops and additional
  * CTRE motor controllers that follow it.
  *
- * @param <T> the type of the motor controller on which the loop is run
  * @author Yoel Perman Brilliant
  * @see DashboardedSubsystem
- * @see SmartMotorControllerSubsystem
+ * @see SmartMotorControllerGenericSubsystem
  * @see BaseMotorController
  * @see MotorController
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/CTRESmartMotorControllerGenericSubsystem.java:26]: remove

File: src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/CTRESmartMotorControllerDrivetrain.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+        double value;
+        switch (controlMode) {
+            case PERCENT_OUTPUT:
+                value = rightMaster.getMotorOutputPercent();
+                break;
+            case VELOCITY:
+                value = rightMaster.getSelectedSensorVelocity();
+                break;
+            case CURRENT:
+                if (leftMaster instanceof BaseTalon) {
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/CTRESmartMotorControllerDrivetrain.java:220]: right

File: src/main/java/com/spikes2212/command/drivetrains/commands/smartmotorcontrollerdrivetrain/MoveSmartMotorControllerTankDrivetrainTrapezically.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param drivetrain               the {@link SmartMotorControllerTankDrivetrain} this command will run on
+     * @param leftPIDSettings          the left side's loop's PID constants
+     * @param rightPIDSettings         the right side's loop's PID constants
+     * @param feedForwardSettings      the loop's feed forward gains
+     * @param leftSetpoint             the setpoint this command should bring the
+     *                                 {@link SmartMotorControllerTankDrivetrain}'s left side to
+     * @param rightSetpoint            the setpoint this command should bring the
+     *                                 {@link SmartMotorControllerTankDrivetrain}'s right side to
+     * @param trapezoidProfileSettings the trapezoid profile setting
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/commands/smartmotorcontrollerdrivetrain/MoveSmartMotorControllerTankDrivetrainTrapezically.java:37]: settings

File: src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+ */
+public class SparkMaxTankDrivetrain extends TankDrivetrain implements SmartMotorControllerTankDrivetrain {
+
+    /**
+     * The slot on the {@link CANSparkMax} on which the trapezoid profiling configurations are saved.
+     */
+    private static final int TRAPEZOID_SLOT_ID = 0;
+
+    /**
+     * The left {@link CANSparkMax} which run the loops.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java:31]: runs

File: src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+     */
+    protected final CANSparkMax leftMaster;
+
+    /**
+     * The right {@link CANSparkMax} which run the loops.
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java:36]: runs

File: src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    /**
+     * Constructs a new instance of {@link SparkMaxTankDrivetrain}.
+     *
+     * @param namespaceName the name of the drivetrain's namespace
+     * @param leftMaster    the {@link CANSparkMax} which runs the left side's loops
+     * @param leftSlave     an additional {@link CANSparkMax} that follows the left master
+     * @param rightMaster   the {@link CANSparkMax} which runs the right side's loops
+     * @param rightSlave    an additional {@link CANSparkMax}s that follows the right master
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java:77]: remove s after sparkmax

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

**Pull request merged**